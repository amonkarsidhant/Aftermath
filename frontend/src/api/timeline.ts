export interface TimelineEvent {
  source: string;
  timestamp: string;
  description: string;
}

export async function fetchTimelineEvents(
  start?: Date,
  end?: Date,
  providers?: string[]
): Promise<TimelineEvent[]> {
  const params = new URLSearchParams();
  if (start) params.set('start', start.toISOString());
  if (end) params.set('end', end.toISOString());
  if (providers && providers.length) params.set('providers', providers.join(','));

  const res = await fetch(`/timeline?${params.toString()}`);
  if (!res.ok) {
    throw new Error('Failed to fetch timeline events');
  }
  const data = await res.json();
  return data.events as TimelineEvent[];
}
