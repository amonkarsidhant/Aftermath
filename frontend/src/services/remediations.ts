export interface AgendaNotes {
  well: string;
  wrong: string;
  lucky: string;
}

export interface Remediation {
  id: string;
  provider: string;
  description: string;
  url: string;
}

const AGENDA_KEY = 'review_agenda';
const ACTIONS_KEY = 'remediation_actions';

export function loadAgenda(): AgendaNotes {
  const raw = localStorage.getItem(AGENDA_KEY);
  return raw ? JSON.parse(raw) : { well: '', wrong: '', lucky: '' };
}

export function saveAgenda(notes: AgendaNotes): void {
  localStorage.setItem(AGENDA_KEY, JSON.stringify(notes));
}

export function loadActions(): Remediation[] {
  const raw = localStorage.getItem(ACTIONS_KEY);
  return raw ? JSON.parse(raw) : [];
}

function saveActions(actions: Remediation[]): void {
  localStorage.setItem(ACTIONS_KEY, JSON.stringify(actions));
}

export async function createAction(
  description: string,
  provider: 'jira' | 'servicenow'
): Promise<Remediation> {
  const res = await fetch('/api/actions', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ description, provider }),
  });
  const data = await res.json();
  const action: Remediation = {
    id: data.id,
    url: data.url,
    description,
    provider,
  };
  const actions = loadActions();
  actions.push(action);
  saveActions(actions);
  return action;
}
