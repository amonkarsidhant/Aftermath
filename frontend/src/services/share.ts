export async function requestShareToken(
  type: 'incident' | 'postmortem',
  id: number | string
): Promise<string> {
  const res = await fetch(`/share/${type}/${id}`, { method: 'POST' });
  if (!res.ok) {
    throw new Error('failed to get share token');
  }
  const data = await res.json();
  return data.token as string;
}

export async function getIncidentShareUrl(id: number | string): Promise<string> {
  const token = await requestShareToken('incident', id);
  return `${window.location.origin}/incidents/${id}?token=${token}`;
}

export async function getPostmortemShareUrl(
  id: number | string
): Promise<string> {
  const token = await requestShareToken('postmortem', id);
  return `${window.location.origin}/postmortems/${id}?token=${token}`;
}

export function getShareTokenFromUrl(): string | null {
  const params = new URLSearchParams(window.location.search);
  return params.get('token');
}

export async function fetchWithShare(
  input: RequestInfo,
  init: RequestInit = {}
) {
  const token = getShareTokenFromUrl();
  const headers = token
    ? { ...(init.headers || {}), Authorization: `Bearer ${token}` }
    : init.headers;
  return fetch(input, { ...init, headers });
}
