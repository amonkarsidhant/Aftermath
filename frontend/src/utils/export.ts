import { jsPDF } from 'jspdf';
import html2canvas from 'html2canvas';
import Papa from 'papaparse';

export async function exportElementToPDF(element: HTMLElement, filename: string) {
  const canvas = await html2canvas(element);
  const imgData = canvas.toDataURL('image/png');
  const pdf = new jsPDF('l', 'pt', [canvas.width, canvas.height]);
  pdf.addImage(imgData, 'PNG', 0, 0, canvas.width, canvas.height);
  pdf.save(filename);
}

export async function exportElementToPNG(element: HTMLElement, filename: string) {
  const canvas = await html2canvas(element);
  const dataUrl = canvas.toDataURL('image/png');
  const link = document.createElement('a');
  link.href = dataUrl;
  link.download = filename;
  link.click();
}

export function exportToCSV(data: object[], filename: string) {
  const csv = Papa.unparse(data);
  const blob = new Blob([csv], { type: 'text/csv;charset=utf-8;' });
  const url = URL.createObjectURL(blob);
  const link = document.createElement('a');
  link.href = url;
  link.download = filename;
  link.click();
  URL.revokeObjectURL(url);
}
