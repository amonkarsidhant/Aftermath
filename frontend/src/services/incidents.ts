import type { Incident } from '../types';
import { config } from '../utils/config';
import mockIncidents from '../utils/mock/incidents.json';

export async function fetchIncidents(): Promise<Incident[]> {
  if (config.useMockData) {
    return mockIncidents as Incident[];
  }
  const res = await fetch('/incidents');
  if (!res.ok) {
    throw new Error('failed to fetch incidents');
  }
  const data = await res.json();
  return data.incidents as Incident[];
}
