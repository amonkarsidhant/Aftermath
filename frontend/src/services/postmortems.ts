import type { Postmortem } from '../types';
import { config } from '../utils/config';
import mockPostmortems from '../utils/mock/postmortems.json';
import { fetchWithShare } from './share';

export async function searchPostmortems(q: string): Promise<Postmortem[]> {
  if (config.useMockData) {
    const normalized = q.toLowerCase();
    return (mockPostmortems as Postmortem[]).filter(
      (pm) =>
        pm.title.toLowerCase().includes(normalized) ||
        pm.summary.toLowerCase().includes(normalized) ||
        pm.tags.some((t) => t.toLowerCase().includes(normalized))
    );
  }
  const res = await fetchWithShare(
    `/postmortems/search?q=${encodeURIComponent(q)}`
  );
  if (!res.ok) throw new Error('failed to search');
  const data = await res.json();
  return data.postmortems as Postmortem[];
}
