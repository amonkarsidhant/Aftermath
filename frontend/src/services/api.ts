import { useQuery } from '@tanstack/react-query';
import type { Incident, Postmortem } from '../types';
import { fetchIncidents } from './incidents';
import { searchPostmortems } from './postmortems';

export function useIncidents() {
  return useQuery<Incident[]>({
    queryKey: ['incidents'],
    queryFn: fetchIncidents,
  });
}

export function usePostmortems(
  query: string,
  options?: { enabled?: boolean }
) {
  return useQuery<Postmortem[]>({
    queryKey: ['postmortems', query],
    queryFn: () => searchPostmortems(query),
    ...options,
  });
}
