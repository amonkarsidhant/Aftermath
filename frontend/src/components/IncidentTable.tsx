import React, { useMemo, useState, useRef } from 'react';
import { useTable, useGlobalFilter, Column } from 'react-table';
import type { Incident } from '../types';
import { useIncidents } from '../services/api';
import LoadingSpinner from './LoadingSpinner';
import ErrorMessage from './ErrorMessage';
import { exportElementToPDF, exportToCSV } from '../utils/export';

interface Props {
  severityFilter?: string;
}

export default function IncidentTable({ severityFilter }: Props) {
  const {
    data: incidents = [],
    isLoading,
    isError,
    refetch,
  } = useIncidents();
  const [severity, setSeverity] = useState('');
  const [service, setService] = useState('');
  const [startDate, setStartDate] = useState('');
  const [endDate, setEndDate] = useState('');
  const tableRef = useRef<HTMLTableElement>(null);

  const severityOptions = useMemo(
    () => Array.from(new Set(incidents.map((i) => i.severity))),
    [incidents]
  );
  const serviceOptions = useMemo(
    () => Array.from(new Set(incidents.map((i) => i.service))),
    [incidents]
  );

  const currentSeverity = severityFilter ?? severity;

  const data = useMemo(() => {
    return incidents.filter((i) => {
      if (currentSeverity && i.severity !== currentSeverity) return false;
      if (service && i.service !== service) return false;
      if (startDate && new Date(i.date) < new Date(startDate)) return false;
      if (endDate && new Date(i.date) > new Date(endDate)) return false;
      return true;
    });
  }, [incidents, currentSeverity, service, startDate, endDate]);

  const columns = useMemo<Column<Incident>[]>(
    () => [
      { Header: 'ID', accessor: 'id' },
      { Header: 'Service', accessor: 'service' },
      { Header: 'Severity', accessor: 'severity' },
      { Header: 'Status', accessor: 'status' },
      { Header: 'Date', accessor: 'date' },
    ],
    []
  );

  const tableInstance = useTable<Incident>({ columns, data }, useGlobalFilter);

  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    rows,
    prepareRow,
    state,
    setGlobalFilter,
  } = tableInstance;

  if (isLoading) {
    return <LoadingSpinner />;
  }

  if (isError) {
    return (
      <ErrorMessage
        message="Failed to load incidents"
        onRetry={refetch}
      />
    );
  }

  return (
    <div className="space-y-4">
      <div className="flex flex-wrap gap-2">
        <input
          value={state.globalFilter ?? ''}
          onChange={(e) => setGlobalFilter(e.target.value)}
          placeholder="Search incidents"
          aria-label="Search"
          className="border border-neutral-300 dark:border-neutral-600 rounded p-1 bg-white dark:bg-neutral-800 text-neutral-900 dark:text-neutral-100"
        />
        <select
          value={currentSeverity}
          onChange={(e) =>
            severityFilter === undefined ? setSeverity(e.target.value) : undefined
          }
          aria-label="Severity"
          className="border border-neutral-300 dark:border-neutral-600 rounded p-1 bg-white dark:bg-neutral-800 text-neutral-900 dark:text-neutral-100"
        >
          <option value="">All Severities</option>
          {severityOptions.map((s) => (
            <option key={s} value={s}>
              {s}
            </option>
          ))}
        </select>
        <select
          value={service}
          onChange={(e) => setService(e.target.value)}
          aria-label="Service"
          className="border border-neutral-300 dark:border-neutral-600 rounded p-1 bg-white dark:bg-neutral-800 text-neutral-900 dark:text-neutral-100"
        >
          <option value="">All Services</option>
          {serviceOptions.map((s) => (
            <option key={s} value={s}>
              {s}
            </option>
          ))}
        </select>
        <input
          type="date"
          value={startDate}
          onChange={(e) => setStartDate(e.target.value)}
          aria-label="Start date"
          className="border border-neutral-300 dark:border-neutral-600 rounded p-1 bg-white dark:bg-neutral-800 text-neutral-900 dark:text-neutral-100"
        />
        <input
          type="date"
          value={endDate}
          onChange={(e) => setEndDate(e.target.value)}
          aria-label="End date"
          className="border border-neutral-300 dark:border-neutral-600 rounded p-1 bg-white dark:bg-neutral-800 text-neutral-900 dark:text-neutral-100"
        />
      </div>
      <div className="flex gap-2">
        <button
          onClick={() =>
            tableRef.current && exportElementToPDF(tableRef.current, 'incidents.pdf')
          }
          className="px-2 py-1 border border-neutral-300 dark:border-neutral-600 rounded"
        >
          Export PDF
        </button>
        <button
          onClick={() => exportToCSV(data, 'incidents.csv')}
          className="px-2 py-1 border border-neutral-300 dark:border-neutral-600 rounded"
        >
          Export CSV
        </button>
      </div>
      <div className="overflow-x-auto">
        <table
          ref={tableRef}
          {...getTableProps()}
          className="min-w-full border border-neutral-300 dark:border-neutral-700 bg-white dark:bg-neutral-800 rounded text-xs sm:text-sm"
        >
          <thead>
            {headerGroups.map((headerGroup) => (
              <tr {...headerGroup.getHeaderGroupProps()}>
                {headerGroup.headers.map((column) => (
                  <th
                    {...column.getHeaderProps()}
                    className="border-b border-neutral-300 dark:border-neutral-700 p-2 text-left bg-neutral-100 dark:bg-neutral-700"
                  >
                    {column.render('Header')}
                  </th>
                ))}
              </tr>
            ))}
          </thead>
          <tbody {...getTableBodyProps()}>
            {rows.map((row) => {
              prepareRow(row);
              return (
                <tr
                  {...row.getRowProps()}
                  className="hover:bg-neutral-100 dark:hover:bg-neutral-700"
                >
                  {row.cells.map((cell) => (
                    <td
                      {...cell.getCellProps()}
                      className="p-2 border-b border-neutral-200 dark:border-neutral-700"
                    >
                      {cell.render('Cell')}
                    </td>
                  ))}
                </tr>
              );
            })}
            {rows.length === 0 && (
              <tr>
                <td colSpan={columns.length} className="p-2 text-center">
                  No incidents
                </td>
              </tr>
            )}
          </tbody>
        </table>
      </div>
    </div>
  );
}

