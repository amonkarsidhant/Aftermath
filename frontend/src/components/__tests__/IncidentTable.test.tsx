import { render, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import IncidentTable from '../IncidentTable';
import { useIncidents } from '../../services/api';
import mockIncidents from '../../utils/mock/incidents.json';
import { exportElementToPDF, exportToCSV } from '../../utils/export';
import { getIncidentShareUrl } from '../../services/share';

jest.mock('../../services/api', () => ({ useIncidents: jest.fn() }));
jest.mock('../../utils/export', () => ({
  exportElementToPDF: jest.fn(),
  exportToCSV: jest.fn(),
}));
jest.mock('../../services/share', () => ({ getIncidentShareUrl: jest.fn() }));

describe('IncidentTable', () => {
  beforeEach(() => {
    (useIncidents as jest.Mock).mockReturnValue({
      data: mockIncidents,
      isLoading: false,
      isError: false,
      refetch: jest.fn(),
    });
  });

  it('matches snapshot', () => {
    const { container } = render(<IncidentTable />);
    expect(container).toMatchSnapshot();
  });

  it('renders all columns', () => {
    render(<IncidentTable />);
    ['ID', 'Service', 'Severity', 'Status', 'Date', 'Share'].forEach((col) => {
      expect(screen.getByText(col)).toBeInTheDocument();
    });
  });

  it('filters by severity', async () => {
    const user = userEvent.setup();
    render(<IncidentTable />);
    await screen.findByText('Auth');
    await user.selectOptions(screen.getByLabelText('Severity'), 'SEV-2');
    const table = screen.getByRole('table');
    expect(within(table).getAllByRole('row')).toHaveLength(3);
    expect(within(table).queryByText('SEV-1')).toBeNull();
  });

  it('searches by id', async () => {
    const user = userEvent.setup();
    render(<IncidentTable />);
    await screen.findByText('Auth');
    const search = screen.getByPlaceholderText('Search incidents');
    await user.type(search, '6');
    const table = screen.getByRole('table');
    expect(within(table).getAllByRole('row')).toHaveLength(2);
    expect(within(table).getByText('6')).toBeInTheDocument();
  });

  it('exports data', async () => {
    const user = userEvent.setup();
    render(<IncidentTable />);
    await screen.findByText('Auth');
    await user.click(screen.getByRole('button', { name: /export pdf/i }));
    expect(exportElementToPDF).toHaveBeenCalled();
    await user.click(screen.getByRole('button', { name: /export csv/i }));
    expect(exportToCSV).toHaveBeenCalled();
  });

  it('copies share link', async () => {
    const user = userEvent.setup();
    (getIncidentShareUrl as jest.Mock).mockResolvedValue('link');
    Object.assign(navigator, { clipboard: { writeText: jest.fn() } });
    render(<IncidentTable />);
    await screen.findByText('Auth');
    const btn = screen.getAllByRole('button', { name: /share/i })[0];
    await user.click(btn);
    expect(getIncidentShareUrl).toHaveBeenCalled();
    expect(navigator.clipboard.writeText).toHaveBeenCalledWith('link');
  });
});

