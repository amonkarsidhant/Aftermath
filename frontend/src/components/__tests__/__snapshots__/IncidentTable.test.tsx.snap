exports[`IncidentTable matches snapshot 1`] = `
<div class="space-y-4">
  <div class="flex flex-wrap gap-2">
    <input value="" placeholder="Search incidents" aria-label="Search" class="border border-neutral-300 dark:border-neutral-600 rounded p-1 bg-white dark:bg-neutral-800 text-neutral-900 dark:text-neutral-100" />
    <select aria-label="Severity" class="border border-neutral-300 dark:border-neutral-600 rounded p-1 bg-white dark:bg-neutral-800 text-neutral-900 dark:text-neutral-100">
      <option value="">All Severities</option>
      <option value="SEV-1">SEV-1</option>
      <option value="SEV-2">SEV-2</option>
      <option value="SEV-3">SEV-3</option>
    </select>
    <select aria-label="Service" class="border border-neutral-300 dark:border-neutral-600 rounded p-1 bg-white dark:bg-neutral-800 text-neutral-900 dark:text-neutral-100">
      <option value="">All Services</option>
      <option value="Auth">Auth</option>
      <option value="Payments">Payments</option>
      <option value="Search">Search</option>
      <option value="Orders">Orders</option>
    </select>
    <input type="date" value="" aria-label="Start date" class="border border-neutral-300 dark:border-neutral-600 rounded p-1 bg-white dark:bg-neutral-800 text-neutral-900 dark:text-neutral-100" />
    <input type="date" value="" aria-label="End date" class="border border-neutral-300 dark:border-neutral-600 rounded p-1 bg-white dark:bg-neutral-800 text-neutral-900 dark:text-neutral-100" />
  </div>
  <div class="flex gap-2">
    <button class="px-2 py-1 border border-neutral-300 dark:border-neutral-600 rounded">Export PDF</button>
    <button class="px-2 py-1 border border-neutral-300 dark:border-neutral-600 rounded">Export CSV</button>
  </div>
  <div class="overflow-x-auto">
    <table role="table" class="min-w-full border border-neutral-300 dark:border-neutral-700 bg-white dark:bg-neutral-800 rounded text-xs sm:text-sm">
      <thead>
        <tr>
          <th class="border-b border-neutral-300 dark:border-neutral-700 p-2 text-left bg-neutral-100 dark:bg-neutral-700">ID</th>
          <th class="border-b border-neutral-300 dark:border-neutral-700 p-2 text-left bg-neutral-100 dark:bg-neutral-700">Service</th>
          <th class="border-b border-neutral-300 dark:border-neutral-700 p-2 text-left bg-neutral-100 dark:bg-neutral-700">Severity</th>
          <th class="border-b border-neutral-300 dark:border-neutral-700 p-2 text-left bg-neutral-100 dark:bg-neutral-700">Status</th>
          <th class="border-b border-neutral-300 dark:border-neutral-700 p-2 text-left bg-neutral-100 dark:bg-neutral-700">Date</th>
          <th class="border-b border-neutral-300 dark:border-neutral-700 p-2 text-left bg-neutral-100 dark:bg-neutral-700">Share</th>
        </tr>
      </thead>
      <tbody>
        <tr class="hover:bg-neutral-100 dark:hover:bg-neutral-700">
          <td class="p-2">1</td>
          <td class="p-2">Auth</td>
          <td class="p-2">SEV-1</td>
          <td class="p-2">Resolved</td>
          <td class="p-2">2024-01-01</td>
          <td class="p-2"><button type="button" class="underline text-primary">Share</button></td>
        </tr>
        <tr class="hover:bg-neutral-100 dark:hover:bg-neutral-700">
          <td class="p-2">2</td>
          <td class="p-2">Payments</td>
          <td class="p-2">SEV-2</td>
          <td class="p-2">Open</td>
          <td class="p-2">2024-01-15</td>
          <td class="p-2"><button type="button" class="underline text-primary">Share</button></td>
        </tr>
        <tr class="hover:bg-neutral-100 dark:hover:bg-neutral-700">
          <td class="p-2">3</td>
          <td class="p-2">Search</td>
          <td class="p-2">SEV-3</td>
          <td class="p-2">In Progress</td>
          <td class="p-2">2024-02-10</td>
          <td class="p-2"><button type="button" class="underline text-primary">Share</button></td>
        </tr>
        <tr class="hover:bg-neutral-100 dark:hover:bg-neutral-700">
          <td class="p-2">4</td>
          <td class="p-2">Auth</td>
          <td class="p-2">SEV-3</td>
          <td class="p-2">Resolved</td>
          <td class="p-2">2024-03-05</td>
          <td class="p-2"><button type="button" class="underline text-primary">Share</button></td>
        </tr>
        <tr class="hover:bg-neutral-100 dark:hover:bg-neutral-700">
          <td class="p-2">5</td>
          <td class="p-2">Orders</td>
          <td class="p-2">SEV-2</td>
          <td class="p-2">Open</td>
          <td class="p-2">2024-04-20</td>
          <td class="p-2"><button type="button" class="underline text-primary">Share</button></td>
        </tr>
        <tr class="hover:bg-neutral-100 dark:hover:bg-neutral-700">
          <td class="p-2">6</td>
          <td class="p-2">Payments</td>
          <td class="p-2">SEV-1</td>
          <td class="p-2">Resolved</td>
          <td class="p-2">2024-05-18</td>
          <td class="p-2"><button type="button" class="underline text-primary">Share</button></td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
